### PR TITLE
Sample SPA HTML app with 3 libraries: AppAuth logout, PKCE, userInfo and AppAuth-JS

### PR DESCRIPTION
**Pull request by ChinthakaSenanayaka**
## Automation tests
 - Unit tests 
   > Code coverage information N/A
 - Integration tests
   > Details about the test cases and coverage N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? N/A
 - Ran FindSecurityBugs plugin and verified report? N/A
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? N/A

### CODE REVIEW

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/package.json [hunk 1]
```diff
@@ -9,7 +9,13 @@
   "author": "Chinthaka Senanayaka",
   "license": "ISC",
   "dependencies": {
-    "@openid/appauth": "file:///home/chinthaka/projects/appauth-lib-dev/AppAuth-JS",
+    "@openid/appauth": "file:///home/chinthaka/projects/identity-client-lib-appauth/AppAuth-JS",
+    "@wso2/appauth-ext-logout": "file:../../../../identity-client-lib-appauth/AppAuth-JS-Ext-Logout",
+    "@wso2/appauth-ext-pkce": "file:///home/chinthaka/projects/identity-client-lib-appauth/AppAuth-JS-Ext-PKCE",
```
Comment (ruwanta) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/package.json]: Some locations are hard-coded

Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/package.json]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 2]
```diff
@@ -32,20 +32,48 @@ var appConfigs = {
...
+  var appArray = [app, appLogout, appPKCE, appUserInfo];
+
+  app.init(authCompletionCallback);
+  appLogout.init(logoutCompletionCallback);
+  appPKCE.init(authCompletionCallback);
+  // There is no init function in AppUserInfo class
 
   // check for authorization response if available.
-  application.checkForAuthorizationResponse();
+  if(appConfigs.flowType == "IMPLICIT") {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: format the source.

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 3]
```diff
@@ -32,20 +32,48 @@ var appConfigs = {
...
+    appPKCE.checkForAuthorizationResponse();
+  }
+  appLogout.checkForAuthorizationResponse();
 
-  return application;
+  return appArray;
 }
+
+var makeAuthorizationRequest = function () {
+  if(appConfigs.flowType == "IMPLICIT") {
```
Comment (ayshsandu) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: Move grant types to constants.

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 4]
```diff
@@ -32,20 +32,48 @@ var appConfigs = {
...
+    app.makeAuthorizationRequest();
+  } else if (appConfigs.flowType == "PKCE") {
+    appPKCE.makeAuthorizationRequest();
+  }
+}
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: add an empty line at the end of the file.

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js [hunk 5]
```diff
@@ -33,14 +35,23 @@ class Home extends Component {
   componentDidMount() {
 
     // Redirect completion callback method execution for authorization completion callback and end session (logout) completion callabck.
-    this.application.checkForAuthorizationResponse();
+    if(this.app.getConfiguration().flowType == "IMPLICIT") {
```
Comment (ayshsandu) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: move grant types to constants

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js [hunk 6]
```diff
@@ -602,6 +528,194 @@ exports.cryptoGenerateRandom = function (sizeInBytes) {
...
+            return result;
+        };
+    }
+    return descriptor;
+}
+
+},{"./flags":8}],10:[function(require,module,exports){
+"use strict";
+/*
+ * Copyright 2017 Google Inc.
```
Comment (ayshsandu) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js]: Incorrect copy-right

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js [hunk 7]
```diff
@@ -0,0 +1,1917 @@
...
+        return encoded.join('&');
+    };
+    return BasicQueryStringUtils;
+}());
+exports.BasicQueryStringUtils = BasicQueryStringUtils;
+
+},{}],17:[function(require,module,exports){
+"use strict";
+/*
+ * Copyright 2017 Google Inc.
```
Comment (ayshsandu) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js]: Fix copy right

**Comment by thanujalk** [comment]
Please add a comment to identify generated sources.

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js [hunk 8]
```diff
@@ -0,0 +1,1917 @@
+(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
```
Comment (ChinthakaSenanayaka) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js]: Generated file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js (hunk 7 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_logout.js]: This is generated source file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_pkce.js [hunk 9]
```diff
@@ -0,0 +1,4374 @@
+(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
```
Comment (ChinthakaSenanayaka) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_pkce.js]: This is generated source file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_userinfo.js [hunk 10]
```diff
@@ -0,0 +1,1577 @@
+(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
```
Comment (ChinthakaSenanayaka) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/appauth_js_ext_userinfo.js]: This is generated source file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js [hunk 11]
```diff
@@ -20,15 +20,9 @@ var authorization_request_handler_1 = require("../authorization_request_handler"
 var authorization_service_configuration_1 = require("../authorization_service_configuration");
```
Comment (ChinthakaSenanayaka) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js]: This is generated source file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js (hunk 6 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/bundle.js]: This is generated source file

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 3 above)
Comment (thanujalk) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: Format the line

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Base.js [hunk 12]
```diff
@@ -43,11 +46,16 @@ class Base extends Component {
             'postLogoutRedirectUri': 'http://localhost:7777/' // Specify the post logout redirect URL to land back after logout redirect to WSO2-IAM Server logout page
           };
       
-        this.application = new App(appConfigs);
+        this.app = new App(appConfigs);
+        this.appLogout = new AppLogout(this.app, appConfigs.postLogoutRedirectUri, appConfigs.clientId);
+        this.appPKCE = new AppPKCE(this.app, appConfigs.clientId, appConfigs.clientSecret, appConfigs.redirectUri, appConfigs.scope, appConfigs.tokenUrl);
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Base.js]: format this line. Length show be <=120

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 13]
```diff
@@ -32,20 +32,48 @@ var appConfigs = {
...
+  appLogout.checkForAuthorizationResponse();
 
-  return application;
+  return appArray;
 }
+
+var makeAuthorizationRequest = function () {
+  if(appConfigs.flowType == "IMPLICIT") {
+    app.makeAuthorizationRequest();
+  } else if (appConfigs.flowType == "PKCE") {
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: Yes, only IMPLICIT is applicable to HTML SPA apps but added only PKCE (Auth code) for extended security

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 3 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: added const and formatted line

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 2 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 4 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js (hunk 5 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: added const and formatted

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Base.js (hunk 12 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Base.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 14]
```diff
@@ -32,20 +32,48 @@ var appConfigs = {
...
+var appUserInfo = null;
+
 var init = function (authCompletionCallback, logoutCompletionCallback) {
-  var application = new App(appConfigs);
 
-  application.init(authCompletionCallback, logoutCompletionCallback);
+  // Remove any object initialization of you are not using and remove the respective js library file in js folder.
+  app = new App(appConfigs);
+  appLogout = new AppLogout(app, appConfigs.postLogoutRedirectUri, appConfigs.clientId);
+  appPKCE = new AppPKCE(app, appConfigs.clientId, appConfigs.clientSecret, appConfigs.redirectUri, appConfigs.scope, appConfigs.tokenUrl);
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js [hunk 15]
```diff
@@ -33,14 +35,23 @@ class Home extends Component {
...
+    }
+    this.appLogout.checkForAuthorizationResponse();
     
   }
 
   loginClick() {
 
     // Execute OIDC authorize requests against WSO2 IS server
-    this.application.makeAuthorizationRequest();
+    if(this.app.getConfiguration().flowType == "IMPLICIT") {
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js [hunk 16]
```diff
@@ -49,16 +50,21 @@ class Profile extends Component {
...
         // Execute OIDC end session (logout) requests against WSO2 IS server
-        this.application.makeLogoutRequest();
+        this.appLogout.makeLogoutRequest();
     }
 
     componentWillMount() {
 
         // Redirect completion callback method execution for authorization completion callback and end session (logout) completion callabck.
-        this.application.checkForAuthorizationResponse().then(this.checkUserLoggedIn(this));
+        if(this.app.getConfiguration().flowType == "IMPLICIT") {
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 17]
```diff
@@ -25,27 +25,64 @@
...
+  var appArray = [app, appLogout, appPKCE, appUserInfo];
+
+  app.init(authCompletionCallback);
+  appLogout.init(logoutCompletionCallback);
+  appPKCE.init(authCompletionCallback);
+  // There is no init function in AppUserInfo class
 
   // check for authorization response if available.
-  application.checkForAuthorizationResponse();
+  if (appConfigs.flowType == FLOW_TYPE_IMPLICIT) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: bring the static value to the front,

FLOW_TYPE_IMPLICIT == appConfigs.flowType

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 18]
```diff
@@ -25,27 +25,64 @@
...
+    app.checkForAuthorizationResponse();
+  } else if (appConfigs.flowType == FLOW_TYPE_PKCE) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: bring the static value to the front.

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 19]
```diff
@@ -25,27 +25,64 @@
...
+    appPKCE.checkForAuthorizationResponse();
+  }
+  appLogout.checkForAuthorizationResponse();
+
+  return appArray;
+}
 
-  return application;
+var makeAuthorizationRequest = function () {
+  if (appConfigs.flowType == FLOW_TYPE_IMPLICIT) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 20]
```diff
@@ -25,27 +25,64 @@
...
+    app.makeAuthorizationRequest();
+  } else if (appConfigs.flowType == FLOW_TYPE_PKCE) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js [hunk 21]
```diff
@@ -33,14 +37,23 @@ class Home extends Component {
   componentDidMount() {
 
     // Redirect completion callback method execution for authorization completion callback and end session (logout) completion callabck.
-    this.application.checkForAuthorizationResponse();
+    if (this.app.getConfiguration().flowType == FLOW_TYPE_IMPLICIT) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js [hunk 22]
```diff
@@ -33,14 +37,23 @@ class Home extends Component {
...
+    }
+    this.appLogout.checkForAuthorizationResponse();
     
   }
 
   loginClick() {
 
     // Execute OIDC authorize requests against WSO2 IS server
-    this.application.makeAuthorizationRequest();
+    if (this.app.getConfiguration().flowType == FLOW_TYPE_IMPLICIT) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js [hunk 23]
```diff
@@ -33,14 +37,23 @@ class Home extends Component {
...
+      this.app.makeAuthorizationRequest();
+    } else if (this.app.getConfiguration().flowType == FLOW_TYPE_PKCE) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js [hunk 24]
```diff
@@ -49,18 +52,22 @@ class Profile extends Component {
...
         // Execute OIDC end session (logout) requests against WSO2 IS server
-        this.application.makeLogoutRequest();
+        this.appLogout.makeLogoutRequest();
     }
 
     componentWillMount() {
 
         // Redirect completion callback method execution for authorization completion callback and end session (logout) completion callabck.
-        this.application.checkForAuthorizationResponse().then(this.checkUserLoggedIn(this));
+        if (this.app.getConfiguration().flowType == FLOW_TYPE_IMPLICIT) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js [hunk 25]
```diff
@@ -49,18 +52,22 @@ class Profile extends Component {
...
+            this.app.checkForAuthorizationResponse().then(this.checkUserLoggedIn(this));
+        } else if (this.app.getConfiguration().flowType == FLOW_TYPE_PKCE) {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js]: bring the static value to the front

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js [hunk 26]
```diff
@@ -25,27 +25,64 @@
 */
 
 var appConfigs = {
```
Comment (thanujalk) [inline oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: Don't we need to add WSO2 license as well?

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 17 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 18 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 19 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 20 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js (hunk 21 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js (hunk 22 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js (hunk 23 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Home.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js (hunk 25 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_React_app/app/Profile.js]: done

File: oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js (hunk 26 above)
Comment (ChinthakaSenanayaka) [reply oidc-client-app-samples/browser-client-app-sample/SPA_HTML_app/js/app.js]: It is already there, you have to collapse the no change content

**Pull request merged**